src/main/java/io/vertx/ext/mongo/impl/MongoClientImpl.java

```diff
@@ -905,8 +907,55 @@ private AggregatePublisher<JsonObject> doAggregate(final String collection, fina
     return aggregate;
   }
 
+
+
+  JsonArray deepEncodeKeyWhenUseObjectId(JsonArray arr) {
+    if(!useObjectId) return arr;
+
+    JsonArray newArr = new JsonArray(new ArrayList<>(arr.size()));
+
+    for (Object item : arr) {
+      if (item instanceof JsonArray) {
+        newArr.add(deepEncodeKeyWhenUseObjectId((JsonArray) item));
+      } else if (item instanceof JsonObject) {
+        newArr.add(deepEncodeKeyWhenUseObjectId((JsonObject) item));
+      } else {
+        newArr.add(item);
+      }
+      // we don't handle cases that value instanceof Map or List
+    }
+
+    return newArr;
+  }
+
+  JsonObject deepEncodeKeyWhenUseObjectId(JsonObject json) {
+    if(!useObjectId) return json;
+
+    JsonObject newJson = new JsonObject(new LinkedHashMap<>(json.size()));
+
+    for (Map.Entry<String, Object> entry : json) {
+      String key = entry.getKey();
+      Object value = entry.getValue();
+      if (key.equals(ID_FIELD)
+          && value instanceof String
+          && ObjectId.isValid((String) value)) {
+        newJson.put(key, new JsonObject().put(JsonObjectCodec.OID_FIELD, value));
+      } else if (value instanceof JsonObject) {
+        newJson.put(key, deepEncodeKeyWhenUseObjectId((JsonObject) value));
+      } else if (value instanceof JsonArray) {
+        newJson.put(key, deepEncodeKeyWhenUseObjectId((JsonArray) value));
+      } else {
+        newJson.put(key, value);
+      }
+      // we don't handle cases that value instanceof Map or List
+    }
+
+    return newJson;
+  }
+
   JsonObject encodeKeyWhenUseObjectId(JsonObject json) {
```

Comment:
Are there any method still calling `encodeKeyWhenUseObjectId`? Why?

Reply:
It is when we reading data from MongoDB, we do not use the `deepEncodeKeyWhenUseObjectId` method to convert the nested `_id`, because if there are such keys exist, it should what the document is.
So when we query, we convert nested `_id` to a JsonObject, but when we read data from MongoDB, we do not need, or say, we should not to do this.

Reply:
Thanks, that makes sense